noir_stdlib/src/hash/sha256.nr

```diff
@@ -169,99 +169,352 @@ unconstrained fn build_msg_block<let N: u32>(
     // We want to check against a constant zero, but if it does not come from the circuit inputs
     // or return values the compiler check will issue a warning.
     let zero = msg_block[0] - msg_block[0];
 
-    for i in 0..BLOCK_SIZE {
-        if i >= msg_byte_ptr {
+    // First integer which is supposed to be (partially) zero.
+    let mut int_byte_ptr = msg_byte_ptr / INT_SIZE;
+
+    // Check partial zeros.
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        let num_zeros = INT_SIZE - modulo;
+        let mask = if num_zeros == 3 {
+            256 * 256 * 256
+        } else if num_zeros == 2 {
+            256 * 256
+        } else {
+            256
+        };
+        assert_eq(msg_block[int_byte_ptr] % mask, zero);
+        int_byte_ptr = int_byte_ptr + 1;
+    }
+
+    // Check the rest of the items.
+    for i in 0..max_int_byte_ptr {
+        if i >= int_byte_ptr {
             assert_eq(msg_block[i], zero);
         }
     }
 }
 
+// Verify that up to the byte pointer the two blocks are equal.
+// At the byte pointer the new block can be partially zeroed.
+fn verify_msg_block_equals_last(
+    msg_block: MSG_BLOCK,
+    last_block: MSG_BLOCK,
+    mut msg_byte_ptr: BLOCK_BYTE_PTR,
+) {
+    // msg_byte_ptr is the position at which they are no longer have to be the same.
+    // First integer which is supposed to be (partially) zero contains that pointer.
+    let mut int_byte_ptr = msg_byte_ptr / INT_SIZE;
+
+    // Check partial zeros.
+    let modulo = msg_byte_ptr % INT_SIZE;
+    if modulo != 0 {
+        // Reconstruct the partially zero item from the last block.
+        let last_field = last_block[int_byte_ptr];
+        let mut msg_item: u32 = 0;
+        // Reset to where they are still equal.
+        msg_byte_ptr = msg_byte_ptr - modulo;
+        for i in 0..INT_SIZE {
+            msg_item = lshift8(msg_item, 1);
+            if i < modulo {
+                msg_item = msg_item + get_item_byte(last_field, msg_byte_ptr) as u32;
+                msg_byte_ptr = msg_byte_ptr + 1;
+            }
+        }
+        assert_eq(msg_block[int_byte_ptr], msg_item);
+    }
+
+    for i in 0..INT_SIZE_PTR {
+        if i < int_byte_ptr {
+            assert_eq(msg_block[i], last_block[i]);
+        }
+    }
+}
+
+// Apply a function on the block item which the pointer indicates.
+fn update_block_item<Env>(
+    mut msg_block: MSG_BLOCK,
+    msg_byte_ptr: BLOCK_BYTE_PTR,
+    f: fn[Env](u32) -> u32,
+) -> MSG_BLOCK {
+    let i = msg_byte_ptr / INT_SIZE;
+    msg_block[i] = f(msg_block[i]);
+    msg_block
+}
+
+// // Pack the value of byte into an item of an existing message block, shifting the byte according to its position.
+// fn set_block_byte(msg_block: MSG_BLOCK, msg_byte_ptr: BLOCK_BYTE_PTR, msg_byte: u8) -> MSG_BLOCK {
+//     update_block_item(
+//         msg_block,
+//         msg_byte_ptr,
+//         |msg_item| set_item_byte(msg_item, msg_byte_ptr, msg_byte),
+//     )
+// }
```

Comment:
```suggestion
```
more leftover looks like, as well as the func below

Reply:
I know, left them for reference, they are nice. I'll remove, I guess it's bothering everyone.